Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -0,0 +1,393 @@
+ * @parameter range string range to consider for rendering.
+ * @parameter frameSize frame constrains to render the text on.
+ * @parameter glyphDetails pointer to the _DWriteGlyphRunDetails object that contains the glyph run details that was rendeered.
+ */
+static void _callCustomRender(_CTTypesetter* ts, CFRange range, CGRect frameSize, _DWriteGlyphRunDetails* glyphDetails) {
+    ComPtr<IDWriteTextLayout> textLayout = _createDWriteTextLayout(ts, range, frameSize);
+    ComPtr<CustomTargetTextRenderer> textRenderer = Make<CustomTargetTextRenderer>();
+    textLayout->Draw(glyphDetails, textRenderer.Get(), 0, 0);
+}
+
+/**
+ * Helper method to retrieve font fmaly names installed in the system.
+ *
+ * @return Unmutable array of font family name strings that are installed in the system.
+ */
+static NSArray<NSString*>* _dwriteGetFamilyNames() {
+    NSMutableArray<NSString*>* fontFamilyNames = [NSMutableArray<NSString*> array];
+
+    // Get the direct write factory instance
+    ComPtr<IDWriteFactory> dwriteFactory = GetDWriteFactoryInstance();
+
+    // Get the system font collection.
+    ComPtr<IDWriteFontCollection> fontCollection;
+    THROW_IF_FAILED(dwriteFactory->GetSystemFontCollection(&fontCollection));
+
+    // Get the number of font families in the collection.
+    uint32_t count = 0;
+    count = fontCollection->GetFontFamilyCount();
+
+    for (uint32_t i = 0; i < count; ++i) {
+        // Get the font family.
+        ComPtr<IDWriteFontFamily> fontFamily;
+        THROW_IF_FAILED(fontCollection->GetFontFamily(i, &fontFamily));
+
+        // Get a list of localized strings for the family name.
+        ComPtr<IDWriteLocalizedStrings> familyNames;
+        THROW_IF_FAILED(fontFamily->GetFamilyNames(&familyNames));
+
+        [fontFamilyNames addObject:_convertLocalizedStringToNSString(familyNames.Get())];
+    }
+
+    return [fontFamilyNames autorelease];
+}
+
+/**
+ * Helper method to create _CTLine objects given a _CTTypesetter, attributed string range to use and frame size to fit in.
+ *
+ * @parameter ts _CTTypesetter object to use.
+ * @parameter range attributed string range to use.
+ * @parameter frameSize size parameters of the frame to fit the text into.
+ *
+ * @return Unmutable array of _CTLine objects created with the requested parameters.
+ */
+static NSArray<_CTLine*>* _dwriteGetLines(_CTTypesetter* ts, CFRange range, CGRect frameSize) {
+    // Call custom renderer to get all glyph run details
+    _DWriteGlyphRunDetails glyphRunDetails;
+    _callCustomRender(ts, range, frameSize, &glyphRunDetails);
+
+    // Create _CTLine objects from the the obtained glyph run details
+    int numOfGlyphRuns = glyphRunDetails._glyphRuns.size();
+    int i = 0;
+    int j = 0;
+
+    NSMutableArray<_CTLine*>* lines = [NSMutableArray array];
+
+    while (j < numOfGlyphRuns) {
+        _CTLine* line = [[_CTLine new] autorelease];
+        NSMutableArray<_CTRun*>* runs = [[NSMutableArray array] retain];
```

Comment:
>NSMutableArray<_CTRun*>* runs = [[NSMutableArray array] retain]; [](start = 8, length = 64)

I see that this isn't leaking because you use an attach below, but to me it seems a lot more comfortable/less prone to error if you don't retain here, and use an = below